### ISSUE TITLE
Support default properties

### ISSUE THREAD

**Issue by agubler** [dojo/widget-core#621]
**Enhancement**

It could be useful for a widget author to be able to specify default properties that the consumer properties are mixed over ensuring that the `this.properties` type guard reflex defaulted properties so that they are not assumed to be possibly `undefined`.

**Comment by pottedmeat** [comment]
This could be helpful for the dgrid editing functionality where there is a set of properties that should be used across all instances - things like length, style, etc. - but each row has its own value.

**Comment by pottedmeat** [comment]
That's fine, I was just outlining a use case.

**Comment by agubler** [comment]
Of course, I think there are multiple use cases for default properties one of which is dgrid. Just wanted to let you know where I had got with trying to implement the feature... it led me down a rabbit hole of typings that I couldn't get quite right unfortunately.

**Comment by kitsonk** [comment]
This seems to be related: Microsoft/TypeScript#16173

There appears to still be some challenges, though it seems like the problem space is better understood.  We should think about it and try to make sure we have sufficient issues on TypeScript to help solve the problem (or support already existing ones).
